src/main.rs

```diff
@@ -1,103 +1,102 @@
+        .compact()
+        .with_line_number(true);
+    let tracing_subscriber = tracing_subscriber::registry()
+        .with(targets_filter)
+        .with(stdout_layer);
+    tracing::subscriber::set_global_default(tracing_subscriber)
+        .into_diagnostic()
+        .map_err(|err| miette::miette!("setting default subscriber failed: {err:#}"))
+}
+
+async fn run_server(validator: Validator, addr: SocketAddr) -> Result<()> {
+    tracing::info!("Listening for gRPC on {addr}");
     Server::builder()
-        .add_service(ValidatorServiceServer::new(bip300.clone()))
+        .add_service(ValidatorServiceServer::new(validator))
         .serve(addr)
         .map_err(|err| miette!("error in validator server: {err:#}"))
         .await
 }
 
+// TODO: return `Result<!, _>` once `never_type` is stabilized
+async fn wallet_task(mut wallet: Wallet) -> Result<(), miette::Report> {
+    const SYNC_INTERVAL: Duration = Duration::from_secs(15);
+    let mut interval_stream = tokio_stream::wrappers::IntervalStream::new(interval(SYNC_INTERVAL));
+    while let Some(_tick) = interval_stream.next().await {
+        let () = wallet.sync()?;
+    }
+    Ok(())
+}
+
 #[tokio::main]
 async fn main() -> Result<()> {
-    env_logger::init();
-
-    let cli = Arc::new(cli::Config::parse());
-
-    let validator = Validator::new(Path::new("./")).map(Arc::new)?;
-
-    // Takes in data from the blockchain and updates the validator state
-    let run_validator_task = tokio::spawn({
-        log::info!("spawning validator task");
-
-        let validator = Arc::clone(&validator);
-        let cli = Arc::clone(&cli);
-        async move { validator.run(&cli).await }
-    });
-
-    let mut tasks: Vec<tokio::task::JoinHandle<Result<(), miette::Error>>> = Vec::new();
-    tasks.push(run_validator_task);
-
-    let run_validator_server_task = tokio::spawn({
-        log::info!("spawning validator server task");
-
-        let validator = Arc::clone(&validator);
-        let cli = Arc::clone(&cli);
-        async move { run_server(&validator, cli.serve_rpc_addr).await }
-    });
-    tasks.push(run_validator_server_task);
-
-    // "Start" the wallet. We're going to add a server here, and run this in a spawned task.
-    // That requires a bit more:
-    // 1. Proper configuration for connecting the wallet to the blockchain
-    // 2. A server for the wallet
-    //
-    // The point here is to prove that we can conditionally start a task.
-    if cli.enable_wallet {
-        let validator = Arc::clone(&validator);
-        let mut wallet = Wallet::new(&cli, &validator)
+    let cli = cli::Config::parse();
+    set_tracing_subscriber(cli.log_level)?;
+
+    let mainchain_client = rpc_client::create_client(&cli.node_rpc_opts)?;
+    let (err_tx, err_rx) = futures::channel::oneshot::channel();
+    let validator = Validator::new(
+        mainchain_client.clone(),
+        cli.node_zmq_addr_sequence,
+        Path::new("./"),
```

Comment:
should this be configurable through CLI opts?

Reply:
I think it should be, although this PR does not change the existing behavior. This can be made configurable in a future PR.